pkg/api/query/grpc.go

```diff
@@ -101,23 +103,36 @@ func (g *GRPCAPI) Query(request *querypb.QueryRequest, server querypb.Query_Quer
 		query.NoopSeriesStatsReporter,
 	)
 
-	var engine promql.QueryEngine
+	var queryEngine promql.QueryEngine
 	engineParam := request.Engine
 	if engineParam == querypb.EngineType_default {
 		engineParam = g.defaultEngine
 	}
 
 	switch engineParam {
 	case querypb.EngineType_prometheus:
-		engine = g.engineFactory.GetPrometheusEngine()
+		queryEngine = g.engineFactory.GetPrometheusEngine()
 	case querypb.EngineType_thanos:
-		engine = g.engineFactory.GetThanosEngine()
+		queryEngine = g.engineFactory.GetThanosEngine()
 	default:
 		return status.Error(codes.InvalidArgument, "invalid engine parameter")
 	}
-	qry, err := engine.NewInstantQuery(ctx, queryable, promql.NewPrometheusQueryOpts(false, lookbackDelta), request.Query, ts)
-	if err != nil {
-		return err
+
+	var qry promql.Query
+	if distEngine, ok := queryEngine.(engine.DistributedEngine); ok && request.QueryPlan != nil {
```

Comment:
I don't think this will work, the engine that handles the grpc request can also be a local engine. It also assumes the engine will be a value type, but in practice it's really a pointer. 

We can remove this fragility completely by declaring a new interface in this repo:

```
type ThanosEngine interface {
	promql.QueryEngine
	NewRangeQueryFromPlan(ctx context.Context, q storage.Queryable, opts promql.QueryOpts, plan logicalplan.Node, start, end time.Time, interval time.Duration) (promql.Query, error)
	NewInstantQueryFromPlan(ctx context.Context, q storage.Queryable, opts promql.QueryOpts, plan logicalplan.Node, ts time.Time) (promql.Query, error)
}

```
The factory should then return `ThanosEngine` for `GetThanosEngine` and inside `case querypb.EngineType_thanos:` we can call `New{Instant, Range}QueryFromPlan`.

Reply:
This looks cleaner now, thx.